pubsub.go

```diff
@@ -162,33 +182,47 @@ func (p *PubsubValueStore) Subscribe(key string) error {
 		return cmp > 0 || cmp == 0 && src == myID
 	})
 
-	sub, err := p.ps.Subscribe(topic)
-	if err != nil {
+	if err := p.fillTopicInfo(ti, topic); err != nil {
 		return err
 	}
 
 	p.mx.Lock()
-	existingSub, bootstrapped := p.subs[key]
-	if existingSub != nil {
-		p.mx.Unlock()
-		sub.Cancel()
-		return nil
-	}
-
 	ctx, cancel := context.WithCancel(p.ctx)
 	p.cancels[key] = cancel
-
-	p.subs[key] = sub
-	go p.handleSubscription(ctx, sub, key, cancel)
+	go p.handleSubscription(ctx, ti, key, cancel)
 	p.mx.Unlock()
 
 	log.Debugf("PubsubResolve: subscribed to %s", key)
 
-	if !bootstrapped {
-		// TODO: Deal with publish then resolve case? Cancel behaviour changes.
-		go bootstrapPubsub(ctx, p.cr, p.host, topic)
+	return nil
+}
+
+func (p *PubsubValueStore) fillTopicInfo(ti *topicInfo, topic string) error {
+	t, err := p.ps.Join(topic)
+	if err != nil {
+		return err
+	}
+
+	sub, err := t.Subscribe()
+	if err != nil {
+		if closeErr := t.Close(); closeErr != nil {
+			return multierror.Append(err, closeErr)
+		}
+		return err
 	}
 
+	evts, err := t.EventHandler()
+	if err != nil {
+		sub.Cancel()
```

Comment:
Will this not be automatically closed when we close the handle?

Reply:
It's currently setup so that you can't close the topic handle if there are existing subscriptions or event handlers (i.e. it returns "cannot close topic: outstanding event handlers or subscriptions")

Reply:
Hm. I'd expect to be able to just close the topic and walk away. Could you file an issue? We can fix this later.

Reply:
We use a similar pattern here where you can't call `Close(topic)` on the PubSubValueStore if SearchValue is ongoing. Should I file an issue here as well?

Reply:
These are two very different APIs:

* In `PubSub`, `Topics` handles are _owned_. If I call `Close` on a topic, I'm not affecting any other services. That's why I was so adamant that we either:
  * Only allowing one service to use a topic at a time (current implementation).
  * Make topic handles completely independent.
* In the `PubSubValueStore`, multiple unrelated services might be independently calling `SearchValue`. I don't want some other service to come along and cancel these searches.

Reply:
Issue filed libp2p/go-libp2p-pubsub#237